From c80d4cb0063950e74ec00bbb8868572ead150126 Mon Sep 17 00:00:00 2001
From: Marco Martin <notmart@gmail.com>
Date: Thu, 8 Aug 2024 12:31:35 +0200
Subject: [PATCH] Set a QQmlEngine property on the QApplication

By setting the "__qmlEngine" property to the QApplication,
if multiple QML-based kcms get loaded, all of them will reuse
this same QQmlEngine instance, working around types mismatches

BUG:488965
---
 src/kcmshell/CMakeLists.txt | 1 +
 src/kcmshell/main.cpp       | 3 +++
 2 files changed, 4 insertions(+)

diff --git a/src/kcmshell/CMakeLists.txt b/src/kcmshell/CMakeLists.txt
index 7256632f..378ac8a6 100644
--- a/src/kcmshell/CMakeLists.txt
+++ b/src/kcmshell/CMakeLists.txt
@@ -7,6 +7,7 @@ ecm_mark_nongui_executable(kcmshell6)
 target_link_libraries(kcmshell6
    KF6KCMUtils
    KF6::I18n
+   Qt6::Qml
    kcmutils_logging_STATIC
 )
 
diff --git a/src/kcmshell/main.cpp b/src/kcmshell/main.cpp
index 3026c4ba..fa45f106 100644
--- a/src/kcmshell/main.cpp
+++ b/src/kcmshell/main.cpp
@@ -13,6 +13,7 @@
 #include <QCommandLineParser>
 #include <QDebug>
 #include <QIcon>
+#include <QQmlEngine>
 #include <QRegularExpression>
 #include <QStandardPaths>
 
@@ -188,6 +189,8 @@ int main(int argc, char *argv[])
         return -1;
     }
 
+    // This ensures if there are multiple QML-based kcms loaded, they use a shared engine instance
+    app.setProperty("__qmlEngine", QVariant::fromValue(new QQmlEngine));
     const bool multipleKCMs = metaDataList.size() > 1;
     KPageDialog::FaceType ftype = multipleKCMs ? KPageDialog::List : KPageDialog::Plain;
     auto dlg = new KCMShellMultiDialog(ftype);
-- 
GitLab

